### DISCUSSION TITLE
How to get the action object that is currently running?

### DISCUSSION THREAD

**Discussion by rudiHammad** [axmolengine/axmol#1852] [Q&A] [answered]
Hello,
I have been checking all over the documentation about actions information, and all I can find is  related to the number of running actions. For instance, say I have a Sprite* called playerSprite. I can do something like this:
`    auto totalRunningActionsNumber = actionManager->getNumberOfRunningActionsInTarget(playerSprite);
`
but that's not really helpful because what I need is the current action object to be return. There are also methods such as `actionManager->getNumberOfRunningActionsInTargetByTag_(playerSprite, someTag)`
but that doesn't help either because this is forcing me to pass the tag which I don't necessarily have to know.

The sprite can be running a walking action, a jumping action, an attacking action, etc... In a nutshell all I want is some thing like (and I am making up the name of the method) _playerSprite->getCurrentAction()_, which would be returning the Action object it self. Or even return the tag of the running action. 

The only way I can think of it, is to create an int variable in the player class, and have it update every frame as the action changes. For instance, if I press "W", then I set that variable to 0, if I press "Space" for the jump, I set it to 1, etc... that way I am "manually" updating the tag. But it doesn't look clean.

thanks

**Comment by rh101** [comment, marked as answer]
Multiple actions can run on a target at the exact same time, and each of those actions can be sequenced, being made up of many other actions, so what you are after wouldn't really help you.

There are many ways to do what you're trying to achieve, with a few examples below.  

You can create a sub-class of Sprite, for example `class PlayerSprite public ax::Sprite`, and add the state variables to that, to track what state the player is currently in.  

You can also create a separate `PlayerState` class, such as `class PlayerState : public ax::Ref` (ensure it inherits from `ax::Ref`), and add it as a a user object to the player sprite, via `Node::setUserObject(Ref* userObject)`.  You would then just do this:

```
auto* playerState = dynamic_cast<PlayerState*>(playerSprite->getUserObject());
auto* currentState = playerState->getState();
playerState->setState(PLAYER_RUNNING); // etc. etc.
```

This also allows you to set the current state *inside* the action sequence, such as by using the `CallFunc` action, and accessing the user object of the target (being the player sprite), to set the state on it.

**Comment by rudiHammad** [comment]
Cool. In fact I already had a PlayerStates class doing something similar to what you mention. I didn't know about the Node::setUserObject() though. Thanks, I'll give that a try.
Cheers.
